### DISCUSSION TITLE
Wrong rect of Document when open "bmp" of "png" files by fitz.open()

### DISCUSSION THREAD

**Discussion by Jianpeng-teng** [pymupdf/PyMuPDF#1969] [Q&A] [answered]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
Open bmp or png files (maybe also other kind of pictures) by fitz.open(file_adress) will return a Document with wrong rect. Which is small than what it should be with a ratio of 0.75.

## To Reproduce (mandatory)
import fitz
img1 = fitz.open("E:\\pic\\0.png")  # 0.png is (1024 x 768)
rect = img1[0].rect  # rect is ( 0, 0, 768, 576)
# you may see the     x_zoom == y_zoom == (768/1024) == (576/768) == 0.75
# And it does not relate to the size of the picture. No matter what picture, the zoom is always 0.75

## Your configuration (mandatory)
 - Operating system, potentially version and bitness
 > Window10 (19044.2130)
 - Python version, bitness
 > Python 3.8.5
 - PyMuPDF version, installation method (**wheel** or **generated** from source).
 > 1.20.2 (I tried 1.20.0 and 1.20.1 also. same)
 > installed with <pip install pymupdf>

For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).

**Comment by JorjMcKie** [comment]
This is no bug, but a misconception about how page sizes of Images are computed when opened as a document.

**Comment by JorjMcKie** [comment]
I will transfer the issue to a `Discussion` and answer it there.

**Comment by JorjMcKie** [comment, marked as answer]
Thanks for asking ... this may be confusing, but has a simple explanation.

For example take this image:
Its dimension is 607 x 741 pixels, its DPI is 120 x 120.
![grafik](https://user-images.githubusercontent.com/8290722/196023365-265ef192-955a-44b8-838c-ef7dcc688b2f.png)
When opened as a document, the corresponding page size is:
```python
page.rect
Rect(0.0, 0.0, 364.20001220703125, 444.6000061035156)
```
The computation works as follows, taking `rect.width` as an example:
```
(img width in pixels) / (x-dpi) * (points per inch)  = page.rect.width
      607             /   120   *        72          = 364.2
```
Or conversely:
```python
page.rect.width * 120 / 72
607.0000203450521
# the original image x-pixels
```

**Reply by Jianpeng-teng** [reply]
Wait, I suddenly realized that, so the Rect size will decrease as DPI increases? That's confusing.

**Comment by Jianpeng-teng** [comment]
Sorry, I was not in front of the screen. that is helpful. So actually it’s happened because the default DPI has been set to 96x96?
It is kind of tricky cause when I try to transfer the pdf document to png and do manipulation and transfer them back, I realized the whole document become smaller than before.

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Jorj X. McKie ***@***.***>
发送时间: Sunday, October 16, 2022 5:30:22 PM
收件人: pymupdf/PyMuPDF ***@***.***>
抄送: Jianpeng-teng ***@***.***>; Author ***@***.***>
主题: Re: [pymupdf/PyMuPDF] Wrong rect of Document when open "bmp" of "png" files by fitz.open() (Discussion #1969)


Thanks for asking ... this may be confusing, but has a simple explanation.

For example take this image:
Its dimension is 607 x 741 pixels, its DPI is 120 x 120.
[grafik]<https://user-images.githubusercontent.com/8290722/196023365-265ef192-955a-44b8-838c-ef7dcc688b2f.png>
When opened as a document, the corresponding page size is:

page.rect
Rect(0.0, 0.0, 364.20001220703125, 444.6000061035156)

The computation works as follows, taking rect.width as an example:

(img width in pixels) / (x-dpi) * (points per inch)  = page.rect.width
      607             /   120   *        72          = 364.2


Or conversely:

page.rect.width * 120 / 72
607.0000203450521
# the original image x-pixels

―

**Reply by JorjMcKie** [reply]
> when I try to transfer the pdf document to png

A little unclear what you are trying to do. But you can influence the PNG size of the PDF page's image (the `Pixmap`)  by choosing an appropriate DPI value. You will of course have that problem, if you leave everything to default values.

**Reply by JorjMcKie** [reply]
When making the page pixmap use the dpi parameter: ``pix = page.get_pixmap(dpi=nnn)``. Choose the value `nnn` according to the above formula, so the PNG opened as document again has the original page dimensions.

**Reply by Jianpeng-teng** [reply]
Thank you so much!
Actually, I just want to crop the pdf page to a new size, like,  the margin of some of the pages were too wide. I found the idea on this page: 
https://pymupdf.readthedocs.io/en/latest/recipes-images.html#how-to-create-partial-pixmaps-clips
>trans = fitz.Matrix(zoom_x, zoom_y)
>clip = fitz.Rect(x1,y1,x2,y2)
>pix = page.get_pixmap(matrix=trans, clip=clip)

And then I couldn't find a simple way to make the pixmap directly back to pdf page.  So:
>pix.save(page_number + ".png")
>doc = fitz.open()
>imgdoc = fitz.open(page_number + ".png")
>pdfbytes = imgdoc.convert_to_pdf()  
>imgdoc.close()
>imgpdf = fitz.open("pdf", pdfbytes)
>doc.insert_pdf(imgpdf) 

(By the way, neither insert_page() nor show_pdf_page() works said in this tutorial:
https://pymupdf.readthedocs.io/en/latest/recipes-images.html#how-to-add-images-to-a-pdf-page
I realized that they had been marked as "private":
![image](https://user-images.githubusercontent.com/60206017/196062165-bb9a5988-935e-401d-90d5-d46d953d185b.png)
Also, the same thing happened to new_page() in:
https://pymupdf.readthedocs.io/en/latest/recipes-general.html#how-to-add-pages
![image](https://user-images.githubusercontent.com/60206017/196062571-4b057af6-2b14-4503-ad60-99e83ba616d9.png)
Are they all replaced by:
>insert_pdf()

?

**Comment by JorjMcKie** [comment]
Okay, so all you want is making the visible part of some page smaller?

If this is true, then using pixmaps and images simply is the wrong idea!
Because a page **_Image_** is something very different from the original page - for example example, all the text is gone, and zooming into the page will show single pixels at some point and more of that kind.

All you need to do is adjusting its CropBox via `page.set_cropbox()`.

**Reply by Jianpeng-teng** [reply]
Ohhh nooo. What am I doing....
